### ISSUE TITLE
Get tokens fails because of a captcha

### ISSUE THREAD

**Issue by euro2** [Chouffy/home_assistant_tgtg#80]
### System Health details

-

### Checklist

- [ ] I have enabled debug logging for my installation.
- [ ] I have filled out the issue template to the best of my ability.
- [ ] This issue only contains 1 issue (if you have multiple issues, open one issue for each issue).
- [ ] This issue is not a duplicate issue of currently [previous issues](https://github.com/Chouffy/home_assistant_tgtg/issues?q=is%3Aissue+label%3A%22Bug%22+)..

### Describe the issue

-

### Reproduction steps

run the get tokens py

### Debug logs

```text
python3 tgtg_get_tokens.py 
Type your email linked to your TGTG account: xxx@hotmail.com
Using version 23.4.12
Traceback (most recent call last):
  File "/root/tgtg_get_tokens.py", line 13, in <module>
    tgtgClient.get_credentials()
  File "/usr/local/lib/python3.10/dist-packages/tgtg/__init__.py", line 90, in get_credentials
    self.login()
  File "/usr/local/lib/python3.10/dist-packages/tgtg/__init__.py", line 181, in login
    raise TgtgLoginError(response.status_code, response.content)
tgtg.exceptions.TgtgLoginError: (403, b'{"url":"https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMAdRgF1adAg_gA2WmLtw==&cid=86PDjw6-EpSxHPjvbU43ToKotX~i8HHzU1h2qr4e-FnmR_iSetqOR-6lSkG2eLprcf~cI8nfz7tzYKidv-OTGGfKhYlmg7yQaRlfYwVm8bL-r56HapF5Z-c6uGd1336&referer=https%3A%2F%2Fapptoogoodtogo.com%2Fapi%2Fauth%2Fv3%2FauthByEmail&hash=1D42C2CA6131C526E09F294FE96F94&t=fe&s=35587&e=6b1d5728fb084e8d880c6e0c40adab58986febb54699b2c4d83229fa46fd6df9"}')
```

**Comment by euro2** [comment]
I now tried a fresh docker run to get tokens. Same result. It errors on the captch url. If you paste it in the browser you get an actual captcha.... Maybe too many polls to their server?

**Comment by joshuaferrara** [comment]
maybe a dumb question, but are you running v5.2.0 of `home_assistant_tgtg`? i requested new tokens using the underlying Python library @ its latest version (see their [README.md](https://github.com/ahivert/tgtg-python#use-it)), upgraded from v5.1.0 to v5.2.0 in HACS w/o restarting, updated my HA configuration to use the new tokens, and then restarted everything. after that I'm back in action.

**Comment by tjorim** [comment]
> I now tried a fresh docker run to get tokens. Same result. It errors on the captch url. If you paste it in the browser you get an actual captcha.... Maybe too many polls to their server?

Define 'fresh docker run', did you also do an clean build (as it might have cached some layers. Try this: `docker build --no-cache https://github.com/Chouffy/home_assistant_tgtg.git#main --tag "homeassistant_tgtg_tokens:latest"`.

**Comment by euro2** [comment]
> > I now tried a fresh docker run to get tokens. Same result. It errors on the captch url. If you paste it in the browser you get an actual captcha.... Maybe too many polls to their server?
> 
> Define 'fresh docker run', did you also do an clean build (as it might have cached some layers. Try this: `docker build --no-cache https://github.com/Chouffy/home_assistant_tgtg.git#main --tag "homeassistant_tgtg_tokens:latest"`.

A fresh run is quite vaque isnt it :)... What i ment is that i installed docker in a new LXC container and issued the commands as described in installation 1b:
`docker build https://github.com/Chouffy/home_assistant_tgtg.git#main --tag "homeassistant_tgtg_tokens:latest"
docker run --rm -it homeassistant_tgtg_tokens`

**Comment by euro2** [comment]
New info: the Home assistant log says captcha again:

`Logger: homeassistant.components.sensor
Source: custom_components/tgtg/sensor.py:189
Integration: Sensor (documentation, issues)
First occurred: 10:04:05 (1 occurrences)
Last logged: 10:04:05

Error while setting up tgtg platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 304, in _async_setup_platform
    await asyncio.shield(task)
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/tgtg/sensor.py", line 76, in setup_platform
    add_entities([TGTGSensor(each_item_id)])
  File "/config/custom_components/tgtg/sensor.py", line 96, in __init__
    self.update()
  File "/config/custom_components/tgtg/sensor.py", line 189, in update
    self.tgtg_answer = tgtg_client.get_item(item_id=self.item_id)
  File "/usr/local/lib/python3.10/site-packages/tgtg/__init__.py", line 276, in get_item
    self.login()
  File "/usr/local/lib/python3.10/site-packages/tgtg/__init__.py", line 152, in login
    self._refresh_token()
  File "/usr/local/lib/python3.10/site-packages/tgtg/__init__.py", line 138, in _refresh_token
    raise TgtgAPIError(response.status_code, response.content)
tgtg.exceptions.TgtgAPIError: (403, b'{"url":"https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMAjokilHqJT8cA2WmLtw==&cid=73DGL~8P1WGJMtY3~PkNluG_n8aH7T3hn8JzlPC~DPsg6SGlH8~E4_4Ec9fhe56p7eD6ZZ9T3he0GuJXiMG80dv1_xVtulrutO7riDwH6_5NbE-4dnofL3jIdSs0e8Nr&referer=https%3A%2F%2Fapptoogoodtogo.com%2Fapi%2Fauth%2Fv3%2Ftoken%2Frefresh&hash=1D42C2CA6131C526E09F294FE96F94&t=fe&s=35558&e=1f8a41279d5b89aed8e94e08de3b9b0ba52cb7b17ec8fd7e1699e673dbb10f51"}')
`

So TGTG thinks im spamming? My refresh is 300sec

**Comment by euro2** [comment]
nope. keep getting captcha's again with  `docker run --rm -it homeassistant_tgtg_tokens`
And the one time this morning it worked and i got the tokens, i pasted over the tokens to HA, but then i i had captchas in the Homeassistant tgtg error logs. So i think im flagged as spammer or something. 

On android i cleared cache and storage of the app and started over. Again a captcha before i could continue with email.

**Comment by euro2** [comment]
I hope this hurdle doesnt kill the Integration :(

**Comment by euro2** [comment]
> 
please keep us updated on your progress

**Comment by Alfagek** [comment]
Still no succes. 2 days ago i did get a update from tgtg. And after that it was down again. And again the captcha puzzle in the tgtg app. So didn't do anything with the intergration in ha. It wasn't doing anything (i think) because i did get the error, intergration is not working because of the captcha 403 error. 

Today i started the app, all working. So restart HA en error captcha 403. Opening the app on phone capctha puzzle again :-(. So did that, restart HA en and again same error 403.

Today i disabled tgtg in HA. And wil wait 24 hours, before i turn it on again. See if that wil do something.

A other strange thing i did mention in HA with TGTG. If i go to terminal and i give the command to get tokens. I do get the next error: `➜  tgtg python3 tgtg_get_tokens.py                                                                                                                     
Traceback (most recent call last):                                                                                                                     
  File "/config/tgtg/tgtg_get_tokens.py", line 4, in <module>                                                                                          
    from tgtg import TgtgClient                                                                                                                        
ModuleNotFoundError: No module named 'tgtg'`

When i "reinstall tgtg" in terminal, its working again. I also did had this a few days ago. I can't remeber if this was also before i did have alle the "tgtg errors". Working now with 2023.6.1

**Comment by Alfagek** [comment]
Well i don't have a clue how to handle this. I'll hope that @tjorim has a few suggestions to try.

Before i did start this afternoon, i started the app on my phone: No captcha puzzle or what so ever. 

I did try to get tokens by using "get tokens". After entering mail adres: 403 error.
Then i did see on my phone, a captcha puzzle was there. So did puzzle, closed app en did try again with same result.
Then i removed tgtg install with "pip uninstall tgtg". And used "pip install tgtg", and tgtg 16 was installing again.
Then i did try again to get the token, same 403 error, and no captcha on the app.
I did try to use the "docker" way on a other virtual machine with HA (on Advanced ssh & Web Terminal) (my test system), same result.

So i'm totally lost.
All did on the same network.

**Comment by euro2** [comment]
i think TGTG noticed the 'api-spam' and introduced captchas.... :(

**Comment by Alfagek** [comment]
Update:

I disabled last weekend the config in HA for Tgtg. And only did use the Tgtg app twice this week. And no captcha puzzle. 

So today i did a new attempt: used “get token” and after entering mail adres, no 403 error 😱

then copie all the thing in to my yaml file. And adjust the pull time to 300 seconds. Made the yaml file active again. And restarted HA, no errors 👌🏻. So that did look good. 

After about 15 minutes the first packs did come visible in HA. And that was about 16:00 my time. It’s now 21:00 and it’s still running without a 403 error. 

So it’s looks like it’s going the right way! 
*fingers crossed*

**Comment by Alfagek** [comment]
It was working great with a 5 minute pull. 

But changed the pull this afternoon to 1 minute and I do get the 403 back again. 

So I’m disable the sensor again and will wait 2 days en place it back with a pull of 5 minutes. Better 5 minutes then a not working addon ;-)

**Comment by Strauman** [comment]
> It was working great with a 5 minute pull.
> 
> But changed the pull this afternoon to 1 minute and I do get the 403 back again.
> 
> So I’m disable the sensor again and will wait 2 days en place it back with a pull of 5 minutes. Better 5 minutes then a not working addon ;-)

Did it work?

**Comment by Alfagek** [comment]
> > It was working great with a 5 minute pull.
> > But changed the pull this afternoon to 1 minute and I do get the 403 back again.
> > So I’m disable the sensor again and will wait 2 days en place it back with a pull of 5 minutes. Better 5 minutes then a not working addon ;-)
> 
> Did it work?

Sorry for the late reaction (holiday). But it did work. It’s running now for a long period and have it set to 180 seconds.

**Comment by Strauman** [comment]
Worked for me too. Anything under 300 sec triggers captcha for me.

**Comment by Strauman** [comment]
> A other strange thing i did mention in HA with TGTG. If i go to terminal and i give the command to get tokens. I do get the next error: `➜ tgtg python3 tgtg_get_tokens.py Traceback (most recent call last): File "/config/tgtg/tgtg_get_tokens.py", line 4, in <module> from tgtg import TgtgClient ModuleNotFoundError: No module named 'tgtg'`
> 
> When i "reinstall tgtg" in terminal, its working again. I also did had this a few days ago. I can't remeber if this was also before i did have alle the "tgtg errors". Working now with 2023.6.1

I suspect that could because the folder name Is the same name as the module. So if you rename /config/tgtg/ to /config/toogood/ it might fix it :)

**Comment by lacrima08** [comment]
Hi !

Same for me :

tgtg.exceptions.TgtgLoginError: (403, b'{"url":"https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMA_q5kzSJSGy0AUkJApQ==&cid=27SBqH17faItuFaWGhCRi2ma3Hko3I38uiij~Ny355AO~ps-2_0nj7UiaEdYpkkmop67DWTgGag8CI11_ttaMUP4lYpyNz0eraKaldsvCAD6BlhrQk3cZ1dyCkgG7VVO&referer=https%3A%2F%2Fapptoogoodtogo.com%2Fapi%2Fauth%2Fv3%2FauthByEmail&hash=1D42C2CA6131C526E09F294FE96F94&t=fe&s=35587&e=1c3370640f9bf883ccd168095c0ca7c6ce3e4edfb9bc5b2fd68bcd86791be076"}')

sensors in HA became unavailable

**Comment by tjorim** [comment]
We can't do anything about the captcha but it's probably related to polling too often (scan_interval).
The default we currently recommend is 900 seconds.
I'm also not sure if it is limited by the total calls per day or too many in a short period.

Some users reported similar issues in the upstream library we use: https://github.com/ahivert/tgtg-python/issues/250.
I pushed the change here https://github.com/Chouffy/home_assistant_tgtg/pull/84 and published version 5.3.0: https://github.com/Chouffy/home_assistant_tgtg/releases/tag/5.3.0.
I'll close this issue for now as we don't have any control over it.
